simulator/src/bin/generate-example-screenshots.rs

```diff
@@ -0,0 +1,216 @@
+//! Generate example screenshots for the embedded-graphics documentation
+//!
+//! Run from the workspace root with:
+//!
+//! ```bash
+//! cargo run --bin generate-example-screenshots | rustfmt +nightly --config-path rustfmt.nightly.toml
+//! ```
+//!
+//! Screenshots are output to `target/drawing-ops`.
+
+use embedded_graphics::{pixelcolor, prelude::*};
+use embedded_graphics_simulator::{OutputSettingsBuilder, SimulatorDisplay};
+
+const OUTPUT_BASE: &str = "./target/drawing-ops";
+
+macro_rules! op {
+    ($display:ident, $title:expr, $description:expr, $code:block) => {
+        $display.clear(pixelcolor::Rgb888::BLACK).unwrap();
+
+        (|| {
+            $code;
+
+            Ok::<(), core::convert::Infallible>(())
+        })().unwrap();
+
+        let output_settings = OutputSettingsBuilder::new().scale(2).build();
+
+        let path = format!("{}/{}.png", OUTPUT_BASE, $title);
+        $display
+            .to_image_buffer(&output_settings)
+            .save(&path)
+            .unwrap();
+
+        let screenshot = base64::encode(std::fs::read(&path).expect("Couldn't open screenshot"));
+
+        let docs = stringify!($code)
+                .trim_matches(|c| c == '{' || c == '}')
+                .trim();
+
+        // Note: empty lines must remain between HTML elements and inner Markdown for the Markdown
+        // to be parsed correctly.
+        println!(
+            r#"//! ## {}
+            //!
+            //! {}
+            //!
+            //! <div style="display: flex">
+            //! <img style="width: 128px; height: 128px; display: inline-block; margin-right: 8px;" alt="{} example screenshot" src="data:image/png;base64,{}" />
+            //! <div style="flex-grow: 1;">
+//!
+//! ```rust
+//! # let mut display = embedded_graphics::mock_display::MockDisplay::default();
+//! {}
+//! # Ok::<(), core::convert::Infallible>(())
+//! ```
+//!
+//! </div>
+//! </div>
+//!"#,
+            $title,
+            $description,
+            $title,
+            screenshot,
+            docs.lines().collect::<Vec<_>>().join("\n//! ")
+        );
+    };
+}
+
+fn main() {
+    let output_base = "./target/drawing-ops";
+    std::fs::create_dir_all(output_base).expect("Output directory could not be created");
+
+    let mut display: SimulatorDisplay<pixelcolor::Rgb888> =
+        SimulatorDisplay::new(Size::new(64, 64));
+
+    op!(
+        display,
+        "Draw a single pixel",
```

Comment:
Should we add a note here that this isn't a preferred way of drawing multiple pixels, but users should instead implement their own iters?

If a display driver implements some form of caching for `draw_iter` and doesn't use an internal framebuffer, drawing pixels this way might be much slower that using an iter. But for prototypes or when performance isn't too important drawing individual pixels this way is fine.

Reply:
Yeah that might be a good idea. Can you come up with a sentence or two and I can put that in?

Reply:
I've added a small note to this example so let me know what you think. I'm wary of adding too much text here as I don't want to make the page too long.

Reply:
Sorry I didn't reply earlier. Maybe something like this could work?

```
For cases where many pixels are drawn it is preferable to implement
a custom iterator instead of calling `Pixel::draw` for each pixel, because
some display drivers implement accelerated drawing of iterators.
```

Reply:
SGTM :) I've added that in so this is good for another look over :ok_hand: